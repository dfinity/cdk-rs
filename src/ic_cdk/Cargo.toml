[package]
name = "ic_cdk"
version = "0.1.0"
authors = ["Hans Larsen <hans@larsen.online>"]
edition = "2018"

[dependencies]
candid = "0.2.4"
cfg-if = "0.1.10"
serde = "1.0.110"
wee_alloc = { version = "0.4.5" }
hex = "0.4.2"
crc32fast = "1.2.0"
base32 = "0.4.0"
