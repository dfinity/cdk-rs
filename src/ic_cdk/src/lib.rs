// Use `wee_alloc` as the global allocator.
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

mod api;
mod futures;
mod ic0;
mod printer;
pub mod storage;

pub use api::*;

static mut DONE: bool = false;

/// Setup the stdlib hooks.
pub fn setup() {
    unsafe {
        if DONE {
            return;
        }
        DONE = true;
    }
    printer::hook()
}

/// Block on a promise in a WASM-friendly way (no multithreading!).
pub fn block_on<F: 'static + std::future::Future<Output = ()>>(future: F) {
    futures::block_on(future);
}

/// Format and then print the formatted message
#[cfg(target_arch = "wasm32")]
#[macro_export]
macro_rules! println {
    ($fmt:expr) => (ic_cdk::print(format!($fmt)));
    ($fmt:expr, $($arg:tt)*) => (ic_cdk::print(format!($fmt, $($arg)*)));
}

/// Format and then print the formatted message
#[cfg(not(target_arch = "wasm32"))]
#[macro_export]
macro_rules! println {
    ($fmt:expr) => (std::print!($fmt));
    ($fmt:expr, $($arg:tt)*) => (std::print!($fmt, $($arg)*));
}

/// Format and then print the formatted message
#[cfg(target_arch = "wasm32")]
#[macro_export]
macro_rules! eprintln {
    ($fmt:expr) => (ic_cdk::print(format!($fmt)));
    ($fmt:expr, $($arg:tt)*) => (ic_cdk::print(format!($fmt, $($arg)*)));
}

/// Format and then print the formatted message
#[cfg(not(target_arch = "wasm32"))]
#[macro_export]
macro_rules! eprintln {
    ($fmt:expr) => (std::eprintln!($fmt));
    ($fmt:expr, $($arg:tt)*) => (std::eprintln!($fmt, $($arg)*));
}
