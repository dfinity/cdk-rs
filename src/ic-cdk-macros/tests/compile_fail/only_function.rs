use ic_cdk::query;

#[query]
struct S;

fn main() {}
