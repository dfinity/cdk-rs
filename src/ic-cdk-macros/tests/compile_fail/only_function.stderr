error: #[query] must be above a function.
       expected `fn`
 --> tests/compile_fail/only_function.rs:4:1
  |
4 | struct S;
  | ^^^^^^
